src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,218 @@ namespace NServiceBus
-                    (id, original) => new VersionedSagaEntity(sagaData, original.LockTokenKey, original));
-            });
-            return TaskEx.CompletedTask;
-        }
+                var entry = GetEntry(sagaData, context);
 
-        void ValidateUniqueProperties(SagaCorrelationProperty correlationProperty, IContainSagaData saga)
-        {
-            var sagaType = saga.GetType();
-            var existingSagas = new List<VersionedSagaEntity>();
-            // ReSharper disable once LoopCanBeConvertedToQuery
-            foreach (var s in data)
-            {
-                if (s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id))
+                if (sagas.TryUpdate(sagaData.Id, entry.UpdateTo(sagaData), entry) == false)
                 {
-                    existingSagas.Add(s.Value);
+                    throw new Exception($"InMemorySagaPersister concurrency violation: saga entity Id[{ sagaData.Id }] already saved.");
                 }
-            }
-            var uniqueProperty = sagaType.GetProperty(correlationProperty.Name);
+            });
 
-            if (correlationProperty.Value == null)
-            {
-                var message = $"Cannot store saga with id '{saga.Id}' since the unique property '{uniqueProperty.Name}' has a null value.";
-                throw new InvalidOperationException(message);
-            }
+            return TaskEx.CompletedTask;
+        }
 
-            foreach (var storedSaga in existingSagas)
+        static Entry GetEntry(IContainSagaData sagaData, ContextBag context)
+        {
+            Entry entry;
+            if (context.TryGet(sagaData.Id.ToString(), out entry) == false)
             {
-                var storedSagaPropertyValue = uniqueProperty.GetValue(storedSaga.SagaData, null);
-                if (Equals(correlationProperty.Value, storedSagaPropertyValue))
-                {
-                    var message = $"Cannot store a saga. The saga with id '{storedSaga.SagaData.Id}' already has property '{uniqueProperty.Name}'.";
-                    throw new InvalidOperationException(message);
-                }
+                throw new Exception("The saga should be retrieved with Get method before it's updated");
             }
+            return entry;
         }
 
-        ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
-        ConcurrentDictionary<string, object> lockers = new ConcurrentDictionary<string, object>();
+        static CorrelationId GetKey(Type sagaType, SagaCorrelationProperty correlationProperty)
+        {
+            var propertyName = correlationProperty.Name;
+            var propertyValue = correlationProperty.Value;
+            return GetKey(sagaType, propertyName, propertyValue);
+        }
 
-        class VersionedSagaEntity
+        static CorrelationId GetKey(Type sagaType, string propertyName, object propertyValue)
```

Comment:
Not sure this method brings much value. I think we could just create it inline but I think @timbussmann will disagree ;)